libraries/chain/block_header_state.cpp

```diff
@@ -216,26 +277,18 @@ void finish_next(const block_header_state& prev,
    // proposer policy
    // ---------------
    next_header_state.active_proposer_policy = prev.active_proposer_policy;
+   next_header_state.latest_proposed_proposer_policy = prev.latest_proposed_proposer_policy;
+   next_header_state.latest_pending_proposer_policy = prev.latest_pending_proposer_policy;
 
-   if (!prev.proposer_policies.empty()) {
-      auto it = prev.proposer_policies.begin();
-      // +1 since this is called after the block is built, this will be the active schedule for the next block
-      if (it->first.slot <= next_header_state.header.timestamp.slot + 1) {
-         next_header_state.active_proposer_policy = it->second;
-         next_header_state.proposer_policies = { ++it, prev.proposer_policies.end() };
-      } else {
-         next_header_state.proposer_policies = prev.proposer_policies;
-      }
-   }
+   evaluate_proposer_policies_for_promotion(prev, next_header_state);
 
    std::optional<proposer_policy> new_proposer_policy;
    if (f_ext.new_proposer_policy_diff) {
       new_proposer_policy = prev.get_last_proposed_proposer_policy().apply_diff(*f_ext.new_proposer_policy_diff);
    }
    if (new_proposer_policy) {
       // called when assembling the block
-      next_header_state.proposer_policies[new_proposer_policy->active_time] =
-         std::make_shared<proposer_policy>(std::move(*new_proposer_policy));
+      next_header_state.latest_proposed_proposer_policy = std::make_shared<proposer_policy>(std::move(*new_proposer_policy));
    }
```

Comment:
I don't believe `apply_diff` returns a `std::optional`.
So the `if` at line 289 can be removed and the whole thing simplified to:

```
   if (f_ext.new_proposer_policy_diff) {
      next_header_state.latest_proposed_proposer_policy =
         std::make_shared<proposer_policy>(prev.get_last_proposed_proposer_policy().apply_diff(*f_ext.new_proposer_policy_diff));
   }
```